guest-agent/VfioService/VfioService/MouseHook.cs

```diff
@@ -0,0 +1,105 @@
+using System.Drawing;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+
+namespace VfioService
+{
+    public class MouseHook
+    {
+        private const int WH_MOUSE_LL = 14;
+        private const int WM_MOUSEMOVE = 0x0200;
+
+        private LowLevelMouseProc callback;
+        private IntPtr hookId = IntPtr.Zero;
+        private ClientManager ClientManager;
+
+        public MouseHook(ClientManager clientManager)
+        {
+            ClientManager = clientManager;
+            callback = HookCallback;
+            SetProcessDpiAwareness(ProcessDPIAwareness.ProcessPerMonitorDPIAware);
+            Process curProcess = Process.GetCurrentProcess();
+            ProcessModule curModule = curProcess.MainModule;
+            hookId = SetWindowsHookEx(WH_MOUSE_LL, callback, GetModuleHandle(curModule.ModuleName), 0);
+        }
+
+        ~MouseHook()
+        {
+            UnhookWindowsHookEx(hookId);
+        }
+
+        private IntPtr HookCallback(int nCode, IntPtr wParam, IntPtr lParam)
+        {
+            if (nCode >= 0 && wParam.ToInt32() == WM_MOUSEMOVE)
+            {
+                MSLLHOOKSTRUCT hookStruct = (MSLLHOOKSTRUCT)Marshal.PtrToStructure(lParam, typeof(MSLLHOOKSTRUCT));
+                var inside = false;
+                foreach (var s in Screen.AllScreens)
+                {
+                    // When a border is hit, the x and y coordinates are outside of the screen's bounds.
+                    // If hitting the left border, a negative x value will be returned.
+                    inside |= s.Bounds.Contains(new Point(hookStruct.pt.x, hookStruct.pt.y));
+                }
+                if (!inside)
+                {
+                    lock (ClientManager.WriteLock)
+                    {
+                        ClientManager.SendCommand(CommandOut.MouseEdged);
+                        ClientManager.SendData(BitConverter.GetBytes(hookStruct.pt.x));
+                        ClientManager.SendData(BitConverter.GetBytes(hookStruct.pt.y));
+                    }
+                }
+            }
+            return CallNextHookEx(hookId, nCode, wParam, lParam);
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct POINT
+        {
+            public int x;
+            public int y;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct MSLLHOOKSTRUCT
+        {
+            public POINT pt;
+            public uint mouseData;
+            public uint flags;
+            public uint time;
+            public IntPtr dwExtraInfo;
+        }
+
+        private delegate IntPtr LowLevelMouseProc(int nCode, IntPtr wParam, IntPtr lParam);
+
+        [DllImport("user32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+        private static extern IntPtr SetWindowsHookEx(int idHook,
+            LowLevelMouseProc lpfn, IntPtr hMod, uint dwThreadId);
```

Comment:
According to [Default Marshaling for delegates](https://docs.microsoft.com/en-us/dotnet/framework/interop/default-marshaling-behavior#default-marshaling-for-delegates) the only contract we need to follow si that the reference to `MouseHook` must live as long as the delegate is being called.